week-3.md

```diff
@@ -0,0 +1,82 @@
+
+Anyway, the BIP is at:
+
+ * https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki
+
+You may wish to skim the "Applications" section since we've got week 5 reserved for looking into that in more detail. However most of the complexity there is actually in reading the detailed explanations of how those work in other papers and documents, so just reading the whole BIP this week might be a good approach too.
+
+Jonas Nick wrote a [blog post](https://medium.com/blockstream/reducing-bitcoin-transaction-sizes-with-x-only-pubkeys-f86476af05d7) for the Blockstream blog about x-only pubkey construction that might be interesting. t-bast from group 17 pointed out a [paper on bad nonces with ECDSA](https://eprint.iacr.org/2019/023.pdf) that may be useful background for why deterministic nonce generation is useful when possible; and Leonid Beder from the same group pointed to a [talk by Andrew Poelstra at the Monero Conference on Schnorr signatures](https://www.youtube.com/watch?v=L6KqkrP_nU4) that mentions an interesting construction to ensure proper nonce generation by hardware wallets. The videos from the [optech taproot workshop](https://bitcoinops.org/en/schorr-taproot-workshop/) might be helpful as a slightly gentler introduction to the topic.
+
+If you're interested in looking at implementations, then you may wish to look at any or all of:
+
+ * the [python reference code](https://github.com/sipa/bips/tree/bip-schnorr/bip-schnorr/) (not for production use)
+ * the [jupyter notebooks from the optech taproot workshop](https://bitcoinops.org/en/schorr-taproot-workshop/) (also not for production use)
+ * [libsecp256k1 PR#558](https://github.com/bitcoin-core/secp256k1/pull/558) adds basic support schnorr signature
+ 
+## Misc
+
+The IRC channel for the Q&A sessions is ##taproot-bip-review on the Freenode IRC network. If you've got questions through the week, feel free to ask there -- if someone is around and available they might be able to answer. As well as the [meeting logs](http://www.erisian.com.au/meetbot/taproot-bip-review/2019/) there are ongoing [channel logs](http://www.erisian.com.au/meetbot/taproot-bip-review/2019/) as well, in case you want to see what's been going on there while you've been absent. As mentioned above, there's only one scheduled Q&A session this week: Tue 1900 UTC (America/Europe/Africa).
+
+ * Tue 19 Nov 11:00 -0800 Los Angeles
+ * Tue 19 Nov 14:00 -0500 New York
+ * Tue 19 Nov 19:00 +0000 UTC / London
+ * Tue 19 Nov 20:00 +0100 Paris
+ * Wed 20 Nov 00:30 +0530 Calcutta
+ * Wed 20 Nov 04:00 +0900 Tokyo
+ * Wed 20 Nov 06:00 +1100 Sydney
+
+There's also the dedicated [slack instance](https://bitcoin-review.slack.com/) if you prefer that method of communicating.
+
+## Correspondence
+
+IRC conversation between Pieter Wuille (sipa) and David Harding (harding). Quoted with permission, links added and some minor typographical errors corrected.
```

Comment:
About including this: I don't object to publishing it, but I feel that perhaps it is a distraction.

There are two issues here. The first is that Bech32 was too vague about only having guarantees against substitution errors (it talks about "number of wrong characters", which can probably be interpreted either way). This was a problem with the writeup. But if people assumed that Bech32 did in fact have guarantees against insertion/deletion errors, there would still not really be a problem. They'd be wrong about thinking the chance of missing detection for those is 0, while it is in fact 1 in a billion - but it would be unlikely for another design to offer better detection properties without worsening overall properties.

The other issue is that regardless of the specific language used in BIP173, there was (and should be) I think a very common expectation that any type of error that doesn't have strict guarantees still will be detected with very high probability (BIP173 also states "As the number of errors goes up, the chance converges towards 1 in 2^30"). This specific insertion/deletion error violates that: it has a 1/32 chance of being not detected, and could have been avoided by better review or analysis by someone more familiar with error detection codes.

I feel that including the conversion moves the focus from the second (a serious design mistake) to the first (being vague about a minor and unavoidable issue).

Reply:
@sipa

> Bech32 was too vague about only having guarantees against substitution errors (it talks about "number of wrong characters" and "error affecting N characters", which can probably be interpreted either way).

Just to be clear, I think better context for that second quote from BIP173 is "**any** error affecting N characters" (emphasis different than original).  As explained in the conversation quoted in this PR, I think that "any error" comes across as something that would include insertion/deletion errors.  Additionally, I believe the "any error" line is the first place where error-detection probability is discussed in the document, so it sets the expectation for readers that other parts of the document that mention errors are describing *any* error.

> if people assumed that Bech32 did in fact have guarantees against insertion/deletion errors, there would still not really be a problem. They'd be wrong about thinking the chance of missing detection for those is 0, while it is in fact 1 in a billion

I don't understand your point here.  In my mind, the problem with people operating under false assumptions when reviewing a specification is only partly that they may receive a different result than expected.  It's also that their understanding is incomplete and so they don't have the comprehension necessary to ask good follow up questions.

Perhaps if I had known that there was a small chance of missing the detection of ins/del errors, I would've asked for more information.  For example, even now knowing about this problem, I'm not sure how it affects the probability of detecting up to four errors of *any* type.  E.g., does it only fail in 1-in-a-billion times for any changed length, or is it 4-in-a-billion if I make four length changes, or is it something else?  Maybe I would've asked for a variation on @gmaxwell's chart showing the detection probability of bech32 vs base58check, since it seems to me that bech32 having at least a 1-in-a-billion chance of failing due to insertion/deletion errors means there's no way it can perform better than base58check.

![](https://people.xiph.org/%7Egreg/temp/bch-eff4-big2.png) 

(Chart sourced from [here](https://www.reddit.com/r/Bitcoin/comments/9b62hb/bech32_address_concerns/e50xt5i/), where Maxwell says, "so long as you make errors like a human does-- a few errors rather than just totally random nonsense that happens to match the character set and length-- the 30-bit checksum in Bech32 is much more protective than a longer 32-bit cryptographic hash."   In retrospect, it seems obvious to me that humans will make insertion/deletion errors.)

> I feel that including the conversion moves the focus from the second (a serious design mistake) to the first (being vague about a minor and unavoidable issue).

I think we would've been much more likely to have caught the "serious design mistake" if more people were aware of the "minor and unavoidable issue".  I think that's relevant to the review club because many reviewers (such as myself!) lack the background and experience to catch serious design mistakes, but most reviewers can look for parts of the specification that are "vague" and ask for clarification, which can help experts discover where there might be undiscovered serious issues.

I think the goal of including the quote in this document is to help encourage reviewers by showing them how even just asking for details about vague or ambiguous wording in the specification may ultimately lead to important new insights.  I think that's good, but I'm also fine with the quote being removed if you (or anyone else) still thinks that it's inappropriate.

Reply:
> Just to be clear, I think better context for that second quote from BIP173 is "any error affecting N characters" (emphasis different than original). As explained in the conversation quoted in this PR, I think that "any error" comes across as something that would include insertion/deletion errors. Additionally, I believe the "any error" line is the first place where error-detection probability is discussed in the document, so it sets the expectation for readers that other parts of the document that mention errors are describing any error.

Yes, I agree this is confusing, and should be improved. I'm not dismissing the fact that the document is vague; I'm arguing it shouldn't be conflated with the fact that there was a distinct and much more serious problem.

> For example, even now knowing about this problem, I'm not sure how it affects the probability of detecting up to four errors of any type. E.g., does it only fail in 1-in-a-billion times for any changed length, or is it 4-in-a-billion if I make four length changes, or is it something else?

Bech32 optimizes for up to 4 substitution errors, which are guaranteed to be detected. For any other class of errors - it doesn't matter where or how or how many, the goal was (approximately) 1 in a billion chance of not being detected.

> I think we would've been much more likely to have caught the "serious design mistake" if more people were aware of the "minor and unavoidable issue".

I'm not convinced about that - perhaps even the opposite. The problem was found by someone who was writing tests for Bech32, and including insertion/deletion errors in the mutations he was trying, perhaps exactly because he believed that the spec claims a guarantee would detect them.

Reply:
I still think there's something useful here to raise, but I'm going to skip trying to include it this week

Reply:
I think @ajtowns is making the right decision by yanking the section from this week's newsletter.  To address @sipa's reply:

> I'm not convinced about that - perhaps even the opposite. The problem was found by someone who was writing tests for Bech32, and including insertion/deletion errors in the mutations he was trying, perhaps exactly because he believed that the spec claims a guarantee would detect them.

I agree, it seems quite possible that person wouldn't have written that test, or at least not have investigated further after the first failure, if the test failure didn't break an inferred guarantee.

OTOH, I offer you an alternative:  what if someone had discovered the specification was unclear about different types of errors and requested it be clarified---which I'm sure you would've done.  Then, I think it's possible that person or someone else would've suggested including the witness program length in the data covered by the bech32 checksum.  (This seems to me like a reasonable theory given that ZmnSCPxj [suggested](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2019-November/017448.html) exactly that less than seven hours after he perhaps first learned about the problem from your post,.)  IIUC, that would bump up the reliability of variable-length v1+ bech32 addresses to the "any error" guarantee BIP173 currently implies.  If that change had been proposed while bech32 was still under review and not widely adopted, I think it would've been an easy change to support.

In short, if the "minor and unavoidable issue" had been described explicitly, perhaps as the result of a reviewer being confused by BIP173's claims, I think we might've avoided the "serious design mistake" altogether even without learning about it.  If true, I again think that's useful motivational material to convey to reviewers who are often afraid that they won't be able to make valuable contributions because they aren't as knowledgeable as the original proposers.  Just clarifying ambiguities in BIPs has the guaranteed result of making the document more informative, and a more informative document may reveal problems that would otherwise have gone unnoticed and unaddressed.

Reply:
> Then, I think it's possible that person or someone else would've suggested including the witness program length in the data covered by the bech32 checksum.

We're far into hypotheticals here, but I feel you really misunderstand the issue here.

Ignoring the insert-q-in-penultimate-position issue for a minute, *anything* but up to 4 substitution errors has a 1 in a billion failure to detect chance. Adding the length into the checksummed data would not change this - an insertion or deletion error would still have a 1 in a billion failure to detect chance. Including the length does not magically turn insertions or deletions into substitution errors.

So I think if anyone would have suggested including the length explicitly without having the understanding of the insert-q bug, it would have been dismissed as needless complication.

Yes, adding the length into the checksum is a workaround for the 1/32 failure chance - but there are simpler solutions too (see my last mail to the list).

I of course agree that being clearer in the BIP on the exact guarantees is better, and I'll try to improve it now. But to me it really feels that this unclarity is an unrelated issue.

Reply:
> I feel you really misunderstand the issue here.

That's quite likely.  I'll stop commenting.